### ISSUE TITLE
unexpected filesystem layout for llvm36.rb

### ISSUE THREAD

**Issue by JamesWidman** [Homebrew/homebrew-versions#707]
I'm not sure if this is intentional:

I did: 

    % brew install llvm36 --with-lld --with-lldb --with-python --without-assertions

... and everything is installed to: 

    % ls -l /usr/local/opt/llvm36
    lrwxr-xr-x 1 james admin 22 Mar  9 10:59 /usr/local/opt/llvm36 -> ../Cellar/llvm36/3.6.0

... and that much is fine. However, every file in:

    /usr/local/Cellar/llvm36/3.6.0/bin

... is a symlink to a file in:

    /usr/local/Cellar/llvm36/3.6.0/lib/llvm-3.6/bin/

... for reasons I can't fathom.

So, why all the extra layers?  Why not just make /usr/local/Cellar/llvm36/3.6.0  the root?

To clarify, the directory structure looks like this:

    % ls -l /usr/local/Cellar/llvm36/3.6.0/
    total 12
    -rw-r--r--  1 james admin  470 Mar  9 10:59 INSTALL_RECEIPT.json
    -rw-r--r--  1 james wheel 3281 Mar 12  2014 LICENSE.TXT
    -rw-r--r--  1 james wheel  663 Jan  7 17:07 README.txt
    drwxr-xr-x 55 james admin 1870 Mar  9 10:59 bin
    drwxr-xr-x  4 james admin  136 Mar  9 10:59 lib
    drwxr-xr-x  4 james admin  136 Mar  9 10:59 share
    % ls -l /usr/local/Cellar/llvm36/3.6.0/lib
    total 0
    drwxr-xr-x 9 james admin 306 Mar  9 10:59 llvm-3.6
    drwxr-xr-x 3 james admin 102 Mar  9 10:59 python2.7
    % ls -l /usr/local/Cellar/llvm36/3.6.0/lib/llvm-3.6
    total 0
    drwxr-xr-x  55 james admin 1870 Mar  9 10:58 bin
    drwxr-xr-x   3 james admin  102 Mar  9 10:58 docs
    drwxr-xr-x  10 james admin  340 Mar  9 10:58 include
    drwxr-xr-x 166 james admin 5644 Mar  9 10:58 lib
    drwxr-xr-x   3 james admin  102 Mar  9 10:44 libexec
    drwxr-xr-x   4 james admin  136 Mar  9 10:58 share
    drwxr-xr-x   3 james admin  102 Mar  9 10:59 usr
    % ls -l /usr/local/Cellar/llvm36/3.6.0/bin
    total 212
    lrwxr-xr-x 1 james admin 29 Mar  9 10:59 FileCheck-3.6 -> ../lib/llvm-3.6/bin/FileCheck
    lrwxr-xr-x 1 james admin 28 Mar  9 10:59 bugpoint-3.6 -> ../lib/llvm-3.6/bin/bugpoint
    lrwxr-xr-x 1 james admin 32 Mar  9 10:59 c-index-test-3.6 -> ../lib/llvm-3.6/bin/c-index-test
    lrwxr-xr-x 1 james admin 27 Mar  9 10:59 clang++-3.6 -> ../lib/llvm-3.6/bin/clang++
    lrwxr-xr-x 1 james admin 25 Mar  9 10:59 clang-3.6 -> ../lib/llvm-3.6/bin/clang

**Comment by JamesWidman** [comment]
[I truncated output of the last command (listing files in `/usr/local/Cellar/llvm36/3.6.0/bin`.]

**Comment by JamesWidman** [comment]
This seems reminiscent of homebrew's normal behavior, where installation involves symlinks like:

    /opt/local/bin/foo -> /opt/local/Cellar/package-with-foo/3.1.4/bin/foo
    /opt/local/lib/libfoo.dylib  -> /opt/local/Cellar/package-with-foo/3.1.4/lib/libfoo.dylib

But it doesn't seem right in this case, because `Cellar/llvm36/3.6.0/lib` doesn't contain symlinks to the files in `Cellar/llvm36/3.6.0/lib/llvm-3.6/lib`.

**Comment by JamesWidman** [comment]
(er, /usr/local, not /opt/local)

**Comment by DomT4** [comment]
Yes, I believe this is intentional. The ` versions ` LLVM formulae follow a slightly different layout to Homebrew/homebrew's LLVM formula as the two are designed to coexist peacefully. An LLVM from Versions should never override the LLVM formula from Homebrew/homebrew, as far as I'm aware. Normally such conflicting formulae would be ` keg_only ` but the LLVM's are too useful/in-demand to lock away.

For example, all of the stuff from ` lib/llvm-{ver}/bin ` is symlinked into the ` bin ` with the version number ` -36 ` in this case, appended, to avoid running into Homebrew's core LLVM which is unversioned. Sticking it in ` lib ` means that if people really need the unversioned ` LLVM ` bin links, they're still available at ` HOMEBREW_PREFIX/lib/llvm-3.6/bin `.

I'll ping @mikemcquaid @jacknagel for a consult on what I just said, but that's my understanding of the logic in the formula at least.

**Comment by jacknagel** [comment]
Yes, it's done this way so that you can install mutiple versions in parallel without conflicts.

For example, all of the files in `bin` are symlinks with version suffixes:

```
% ls -l /usr/local/Cellar/llvm36/3.6.0/bin
total 212
lrwxr-xr-x 1 james admin 29 Mar  9 10:59 FileCheck-3.6 -> ../lib/llvm-3.6/bin/FileCheck
lrwxr-xr-x 1 james admin 28 Mar  9 10:59 bugpoint-3.6 -> ../lib/llvm-3.6/bin/bugpoint
lrwxr-xr-x 1 james admin 32 Mar  9 10:59 c-index-test-3.6 -> ../lib/llvm-3.6/bin/c-index-test
lrwxr-xr-x 1 james admin 27 Mar  9 10:59 clang++-3.6 -> ../lib/llvm-3.6/bin/clang++
lrwxr-xr-x 1 james admin 25 Mar  9 10:59 clang-3.6 -> ../lib/llvm-3.6/bin/clang
```

So, not a bug.

**Comment by JamesWidman** [comment]
To @DomT4 , It wouldn't run into Homebrew's core llvm anyway, because the files your'e talking about are all under `Cellar/llvm36`, not `Cellar/llvm`.

To both, I think understand the point of having the symlink names be suffixed with "-${version}"; an upshot is that the symlink:

    /usr/local/bin/clang-3.6 -> ../Cellar/llvm36/3.6.0/bin/clang-3.6 ( -> ../lib/llvm-3.6/bin/clang)

... would not conflict with:

    /usr/local/bin/clang

... right?  (As it happens, formula `llvm` doesn't install `clang` into `/usr/local/bin`, but if it ever does, there won't be a conflict with formula `llvm36`, which *does* install `clang-3.6` into `/usr/local/bin`.)

That much seems clear. 

What confuses me is the redundancy in the full path:  when I see:

    /usr/local/Cellar/llvm36/3.6.0/lib/llvm-3.6

... I'm not sure why, *after* "`3.6.0`", there is a duplicated mention of the version number "3.6".  I guess it allows a namespace for custom patching? E.g.:

    /usr/local/Cellar/llvm36/3.6.0/lib/llvm-3.6     # plain vanilla 3.6.0
    /usr/local/Cellar/llvm36/3.6.0/lib/llvm-3.6-X  # 3.6.0 with my funky X patch 
    /usr/local/Cellar/llvm36/3.6.0/lib/llvm-3.6-Y  # 3.6.0 with my funky Y patch

... where X and Y are names of patches that I've worked on locally.

Is this what you mean by  "multiple versions in parallel without conflicts"?  If so, that makes sense; otherwise I'm still confused. Because then, it seems like you can have conflict-free installations with just:

    /usr/local/Cellar/llvm36/3.6.0/bin/clang  # clang version 3.6.0 (not a symlink)
    /usr/local/Cellar/llvm36/3.6.1/bin/clang  # clang version 3.6.1 (not a symlink)
    /usr/local/Cellar/llvm36/3.6.2/bin/clang  # clang version 3.6.2 (not a symlink)

... and appropriate symlinks, similar to what you have already:

    /usr/local/bin/clang-3.6 -> ../Cellar/llvm36/3.6.2/bin/clang  # rest are unlinked

... Right?  (If you need a particular patchlevel, like 3.6.1 instead of .2, then, as a user, you would need to write "`Cellar/llvm36/3.6.2/`" anyway, and then the symlink doesn't seem to be doing anything that you haven't already done for yourself.)

**Comment by JamesWidman** [comment]
oh... or, did you mean that `homebrew` and `homebrew-versions` might *both* have a formula that puts files under `Cellar/llvm36`?  *That* would be a little surprising (but it would explain things).

**Comment by JamesWidman** [comment]
Oh, and @DomT4 : llvm *is* keg-only. (At the moment, anyway.)

**Comment by DomT4** [comment]
> because the files your'e talking about are all under Cellar/llvm36, not Cellar/llvm.

It was a somewhat bad example, as Homebrew/Homebrew's LLVM is `keg_only`.

A better example would be two versioned LLVM formulae; Whilst the distinct formula-level paths are ` HOMEBREW_CELLAR/llvm35 ` and ` HOMEBREW_CELLAR/llvm36 ` both formulae symlink into ` HOMEBREW_PREFIX `, and that is where the conflict arises. If the prefix was just ` prefix ` and libs were placed directly into ` HOMEBREW_PREFIX/lib ` then you'd hit conflict installing multiple versions.

There are multiple ways to deal with such conflict, ` keg_only`, which is sealing the tool within its own directory and only accessible via the ` opt ` directory, so not available directly in the ` $PATH `, ` conflicts_with ` which means you couldn't have LLVM36 and LLVM35 installed at the same time, and installing to ` lib ` (in this case) and then symlinking the conflict either inside versioned folders or with a binary version suffix.

>  there is a duplicated mention of the version number "3.6"

Because the ` lib/llvm-3.6 ` directory *is* symlinked into ` HOMEBREW_PREFIX/lib ` as ` llvm36 ` - If we just symlinked in the libs directly by setting the top-level ` prefix ` we would create conflict with any other non-keg-only LLVM formulae. The folder versioning prevents that symlink conflict.

> Is this what you mean by "multiple versions in parallel without conflicts"?

No. I mean you can install ` llvm35 ` from Homebrew/homebrew-versions, and ` llvm36 ` Homebrew/homebrew-versions without ever running into a symlink conflict, and so on, because when they are symlinked they're symlinked like this:

` Cellar/LLVM35/lib/llvm35 => HOMEBREW_PREFIX/lib/llvm35 `
` Cellar/LLVM36/lib/llvm36 => HOMEBREW_PREFIX/lib/llvm36 `

Whereas otherwise you'd get something like

` Cellar/LLVM35/lib/libLTO.dylib => HOMEBREW_PREFIX/lib/libLTO.dylib `
` Cellar/LLVM36/lib/libLTO.dylib => HOMEBREW_PREFIX/lib/libLTO.dylib `

**Comment by DomT4** [comment]
> Oh, and @DomT4 : llvm is keg-only. (At the moment, anyway.)

Yes. I made a mistake on the example and should have used two versions formulae to explain my point. I corrected that slip in the above comment.

**Comment by JamesWidman** [comment]
> No. I mean you can install llvm35 from Homebrew/homebrew-versions, and llvm36 Homebrew/homebrew-versions without ever running into a symlink conflict, and so on, because when they are symlinked they're symlinked like this:
>
>     Cellar/LLVM35/lib/llvm35 => HOMEBREW_PREFIX/lib/llvm35
>     Cellar/LLVM36/lib/llvm36 => HOMEBREW_PREFIX/lib/llvm36
>
> Whereas otherwise you'd get something like
>
>     Cellar/LLVM35/lib/libLTO.dylib => HOMEBREW_PREFIX/lib/libLTO.dylib
>     Cellar/LLVM36/lib/libLTO.dylib => HOMEBREW_PREFIX/lib/libLTO.dylib


Ah, I see!

I erroneously expected that, since `HOMEBREW_PREFIX/bin` contains `clang-3.6` (which links to a file, not a directory), then `HOMEBREW_PREFIX/lib` would contain `libLTO-3.6.dylib` (which likewise would link to a file). But instead,  `HOMEBREW_PREFIX/lib` contains `llvm-3.6`, which links to a directory.

Ok, so why is it that:

- `bin/` contains versioned links to *files* but
- `lib/` contains a versioned link to a *directory*?  

(I looked in the docs but couldn't find reference to this.)

It seems like, one way or another, the user has to explicitly write the version number---whether it's in the name of a symlink-to-file or the name of a symlink-to-directory.  But depending on the kind of file (program image, library file, other, etc), the convention used to name the thing differs.

For consistency, would it not be simpler to just have:

    HOMEBREW_PREFIX/opt/llvm36/bin/clang
    HOMEBREW_PREFIX/opt/llvm36/lib/libLTO.dylib

... and then just not install anything under `HOMEBREW_PREFIX/(bin|lib)`?

(I'm not actually proposing a change here; just trying to understand why things are the way they are.)

**Comment by DomT4** [comment]
Because the tools symlinked into ` bin ` know they really exist in ` lib/bin ` and consequently know the libs they need for runtime exist in ` lib/lib `. If you need to use those particular libs outside of using the tools themselves, a simple ` LDFLAG ` or even temporarily prepending your ` $PATH ` with the lib/bin directory will suffice; It's a relatively pleasant compromise between usability and not borking the system. There's also greater risk of cross-contamination against the system the more we symlink.

I'm not disagreeing that ` keg_only ` would in many ways make life easier, but I presume there has been enough demand across the years that the versions formulae remain accessible directly from the ` $PATH `. The current system was the compromise put in place [two years ago](https://github.com/Homebrew/homebrew-versions/commit/1f414bcd47d2f02eb1ca3c2f9401131d20403637) to enable specifically ` keg_only ` not to be enforced on the LLVM formulae.
